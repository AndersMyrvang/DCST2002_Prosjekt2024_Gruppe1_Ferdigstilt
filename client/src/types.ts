// Frontend Types er samlet her

//Brukes i league-service
export type League = {
  id: number;
  name: string;
  country_id: number;
  country_name: string;
  emblem_image_url: string;
  page_id: number;
  content: string;
};

//Brukes i sideHistorikken
export type Page = {
  page_id?: number;
  created_by?: number;
  created_at: Date;
  view_count: number;
};

export type Revisions = {
  revision_id: number;
  content: string;
  revised_by: number;
  revised_at: Date;
  page_id: number;
  username?: string;
};

//Brukes i player-service
export type Player = {
  id: number;
  name: string;
  birth_date: Date;
  height: number;
  country_id: number;
  country_name: string;
  team_name: string;
  team_id: number;
  picture_url: string;
  page_id: number;
  content: string;
  league_id?: number;
  league_name?: string;
};

export type Country = {
  country_name: string;
  flag_image_url: string;
  country_id: number;
};

//Brukes ogs√• i team-service
export type Team = {
  name: string;
  id: number;
  country_id: number;
  country_name: string;
  coach: string;
  league_id: number;
  league_name: string;
  emblem_image_url: string;
  page_id: number;
  content: string;
};

//Brukes i comment-service
export type Comment = {
  comment_id: number;
  user_id: number;
  content: string;
  created_at: Date;
  updated_at: Date;
  page_id: number;
  likes: number;
  profile_image_url?: string;
  username?: string;
};

// Brukes i profile-component
export type UserProfile = {
  username: string;
  email: string;
  photo: string;
  firstLogin: string;
  lastLogin: string;
  is_admin: boolean;
  user_id: number;
};

// Brukes i searchBar-component
export type Suggestion = {
  type: string;
  id?: number;
  name: string;
};

// Brukes i API-respons for brukerdata (backend)
export type PublicUser = {
  username: string;
  email: string | null;
  photo: string | null;
  firstLogin: string;
  lastLogin: string;
  is_admin: boolean;
  user_id: number;
};

export type User = {
  user_id: number;
  google_id: string;
  username: string;
  email?: string | null;
  profile_image_url?: string | null;
  first_login: Date;
  last_login: Date;
  last_logout?: Date | null;
  is_admin: boolean;
};

export type Tag = {
  tag_id: number;
  tag_name: string;
};
