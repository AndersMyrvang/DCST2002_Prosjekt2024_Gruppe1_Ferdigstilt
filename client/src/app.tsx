import React, { useEffect, useState } from 'react';
import { HashRouter as Router, Route } from 'react-router-dom';
import NavBar from './components/navBar-components';
import { PlayerDetails } from './components/player-components';
import { TeamDetails } from './components/team-components';
import { LeagueDetails } from './components/league-components';

// Dummy Home component as placeholder
function Home() {
  return <div>Welcome Home</div>;
}

function App() {
  interface User {
    displayName: string;
  }

  const [user, setUser] = useState<User | null>(null);

  // Fetching user info from /api/current_user when the app loads
  useEffect(() => {
    fetch('/api/current_user')
      .then((res) => res.json())
      .then((data) => {
        setUser(data);
      });
  }, []);

  return (
    <Router>
      {/* Define routes for the page */}
      <Route exact path="/" component={Home} />
      <Route path="/players/:name" component={PlayerDetails} />
      <Route path="/teams/:name" component={TeamDetails} />
      <Route path="/leagues/:name" component={LeagueDetails} />
    </Router>
  );
}

export default App;
