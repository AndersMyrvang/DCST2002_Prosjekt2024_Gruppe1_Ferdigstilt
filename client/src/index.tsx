import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert } from './widgets';
import {
  PlayerList,
  PlayerDetails,
  PlayerEdit,
  PlayerListAll,
  PlayerNew,
} from './components/player-components';
import {
  TeamList,
  TeamDetails,
  TeamEdit,
  TeamListAll,
  TeamNew,
} from './components/team-components';
import { TagList, TagAdd } from './components/tag-component';
import {
  LeagueList,
  LeagueDetails,
  LeagueEdit,
  LeagueListAll,
  LeagueNew,
} from './components/league-components';
import Hamburger from './components/navBar-components';
import Profile from './components/profile-component';

class Menu extends Component {
  render() {
    return <Hamburger user={null} />;
  }
}

class Home extends Component {
  render() {
    return (
      <Card title="Welcome">
        <PlayerList />
        <TeamList />
        <LeagueList />
      </Card>
    );
  }
}

let root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <HashRouter>
      <div>
        <Alert />
        <Menu />
        <Route exact path="/" component={Home} />
        <Route exact path="/menu" component={Hamburger} />

        <Route exact path="/page" component={PlayerList} />
        <Route exact path="/players" component={PlayerListAll} />
        <Route exact path="/player/:id" component={PlayerDetails} />
        <Route exact path="/player/:id/edit" component={PlayerEdit} />
        <Route exact path="/player_id/new" component={PlayerNew} />

        <Route exact path="/page" component={TeamList} />
        <Route exact path="/teams" component={TeamListAll} />
        <Route exact path="/team/:id/" component={TeamDetails} />
        <Route exact path="/team/edit/:id/" component={TeamEdit} />
        <Route exact path="/team_id/new" component={TeamNew} />

        <Route exact path="/page" component={LeagueList} />
        <Route exact path="/leagues" component={LeagueListAll} />
        <Route exact path="/league/:id" component={LeagueDetails} />
        <Route exact path="/league/:id/edit" component={LeagueEdit} />
        <Route exact path="/league_id/new" component={LeagueNew} />

        <Route exact path="/tags" component={TagList} />
        <Route exact path="/tags/new" component={TagAdd} />
        <Route exact path="/profile" component={Profile} />
      </div>
    </HashRouter>,
  );
