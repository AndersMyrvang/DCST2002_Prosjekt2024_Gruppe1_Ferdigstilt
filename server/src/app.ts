import express from 'express';
import playerRouter from './routes/player-router';
import teamRouter from './routes/team-router';
import leagueRouter from './routes/league-router';
import tagRouter from './routes/tag-router';
import commentRouter from './routes/comment-router';
import revisionRouter from './routes/revision-router';
/**
 * Express application.
 */
const app = express();

app.use(express.json());

app.use('/api/v2', playerRouter);
app.use('/api/v2', teamRouter);
app.use('/api/v2', leagueRouter);
app.use('/api/v2', tagRouter);
app.use('/api/v2', commentRouter);
app.use('/api/v2', revisionRouter);

export default app;
