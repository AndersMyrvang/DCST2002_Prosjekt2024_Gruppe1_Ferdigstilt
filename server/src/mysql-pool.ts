import mysql from 'mysql2';
import sqlite3 from 'sqlite3';
const mkdirp = require('mkdirp');

mkdirp.sync('./var/db');

const sqliteDB = new sqlite3.Database('./var/db/sessions.db');

const pool = mysql.createPool({
  host: process.env.MYSQL_HOST,
  user: process.env.MYSQL_USER,
  password: process.env.MYSQL_PASSWORD,
  database: process.env.MYSQL_DATABASE,
  connectionLimit: 1,
  typeCast: (field, next) =>
    field.type == 'TINY' && field.length == 1 ? field.string() == '1' : next(),
});

export { pool, sqliteDB };
