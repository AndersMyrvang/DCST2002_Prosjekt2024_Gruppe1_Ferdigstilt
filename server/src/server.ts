import express from 'express';
import path from 'path';
import session from 'express-session';
import passport from 'passport';
import { Strategy as GoogleStrategy } from 'passport-google-oauth20';
import dotenv from 'dotenv';
import app from './app';
import userRouter from './routes/user-router';
import userService from './services/user-service';
import searchService from './services/search-service';
import playerService from './services/player-service';
import { User } from '../../client/src/types';

dotenv.config();

const PORT = process.env.PORT || 3000;

// Konfigurerer session
app.use(
  session({
    secret: process.env.SESSION_SECRET || 'hemmelignokkel',
    resave: false,
    saveUninitialized: true,
  }),
);


// Initialiserer Passport
app.use(passport.initialize());
app.use(passport.session());

// Google OAuth-strategi
passport.use(
  new GoogleStrategy(
    {
      clientID: process.env.GOOGLE_CLIENT_ID as string,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
      callbackURL: 'http://localhost:3000/auth/google/callback',
    },
    async (accessToken, refreshToken, profile, done) => {
      try {
        const googleId = profile.id;
        const displayName = profile.displayName;
        const email = profile.emails ? profile.emails[0].value : '';
        const photo = profile.photos ? profile.photos[0].value : '';

        const user: User = {
          user_id: 0, 
          google_id: googleId,
          username: displayName,
          email: email,
          profile_image_url: photo,
          first_login: new Date(),
          last_login: new Date(),
          last_logout: null,
          is_admin: false,
        };

        const savedUser = await userService.findOrCreateUserByGoogleId(googleId, user);
        return done(null, { user_id: savedUser.user_id }); 
      } catch (error) {
        console.error('Feil under lagring av brukerdata:', error);
        return done(error, false);
      }
    },
  ),
);

passport.serializeUser((user, done) => {
  done(null, (user as any).user_id);
});

passport.deserializeUser(async (id, done) => {
  try {
    const user = await userService.getUserById(id as number);
    done(null, user);
  } catch (error) {
    done(error, null);
  }
});

// Google autentiseringsruter
app.get('/auth/google', passport.authenticate('google', { scope: ['profile', 'email'] }));
app.get(
  '/auth/google/callback',
  passport.authenticate('google', { failureRedirect: '/' }),
  (req, res) => {
    res.redirect('/');
  },
);

// Rute for å logge ut og oppdatere `last_logout`
app.get('/logout', async (req, res) => {
  if (req.user) {
    const userId = (req.user as any).user_id;

    try {
      await userService.updateLastLogout(userId);
    } catch (error) {
      console.error('Feil ved oppdatering av last_logout:', error);
    }
  }

  req.logout((err) => {
    if (err) {
      console.error('Feil ved utlogging:', err);
      return res.status(500).send('Utlogging mislyktes');
    }
    res.redirect('/');
  });
});

app.use('/api', userRouter);

app.get('/api/search', async (req, res) => {
  const query = req.query.q as string;

  if (!query) {
    return res.status(400).json({ error: 'Søkespørsmål er påkrevd' });
  }

  try {
    const players = await playerService.searchPlayers(query);
    res.json({ players });
  } catch (error) {
    console.error('Feil under søket:', error);
    res.status(500).json({ error: 'Noe gikk galt under søket' });
  }
});

app.get('/api/search/suggestions', async (req, res) => {
  const query = req.query.q as string;

  if (!query || query.length < 3) {
    return res.status(400).json({ error: 'Søkespørsmål må inneholde minst tre tegn' });
  }

  try {
    const suggestions = await searchService.searchMultipleTables(query);
    res.json({ suggestions });
  } catch (error) {
    console.error('Feil under henting av søkeforslag:', error);
    res.status(500).json({ error: 'Noe gikk galt under henting av søkeforslag' });
  }
});

app.use(express.static(path.join(__dirname, '/../../client/public')));

app.listen(PORT, () => {
  console.info(`Server kjører på port ${PORT}`);
});