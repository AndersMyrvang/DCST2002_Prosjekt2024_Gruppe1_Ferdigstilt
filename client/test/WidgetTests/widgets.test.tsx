import React from 'react';
import { shallow, mount } from 'enzyme';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Card, Card2, Card3, Button, Row, Column, Alert, Details } from '../../src/widgets';
import { NavLink } from 'react-router-dom';
import { Action } from 'history';

beforeAll(() => {
  localStorage.setItem('user_id', '1');
});

afterAll(() => {
  localStorage.clear();
});

jest.mock('../../src/services/user-service', () => ({
  getUserById: jest.fn().mockResolvedValue({ user_id: 1, username: 'Test User' }),
}));

describe('Widgets Component Tests', () => {
  describe('Card Components', () => {
    test('Card renders with title and children', () => {
      const wrapper = shallow(
        <Card title="Test Card">
          <p>Card Content</p>
        </Card>,
      );
      expect(wrapper.find('.card-title1').text()).toBe('Test Card');
      expect(wrapper.find('.card-text1').contains(<p>Card Content</p>)).toBeTruthy();
    });

    test('Card2 renders with title and children', () => {
      const wrapper = shallow(
        <Card2 title="Test Card2">
          <p>Card2 Content</p>
        </Card2>,
      );
      expect(wrapper.find('.card-title2').text()).toBe('Test Card2');
      expect(wrapper.find('.card-content2').contains(<p>Card2 Content</p>)).toBeTruthy();
    });

    test('Card3 renders with title and children', () => {
      const wrapper = shallow(
        <Card3 title="Card3 Title">
          <p>Card3 Content</p>
        </Card3>,
      );

      expect(wrapper.find('.card-title').text()).toBe('Card3 Title');

      expect(wrapper.find('.card-content').text()).toBe('Card3 Content');

      expect(wrapper.find('.card').exists()).toBe(true);
      expect(wrapper.find('.card-body').exists()).toBe(true);
    });
  });

  describe('Row and Column Components', () => {
    test('Row renders children properly', () => {
      const wrapper = shallow(
        <Row>
          <p>Row Content</p>
        </Row>,
      );
      expect(wrapper.find('.row').contains(<p>Row Content</p>)).toBeTruthy();
    });

    test('Column renders with width and children', () => {
      const wrapper = shallow(
        <Column width={6}>
          <p>Column Content</p>
        </Column>,
      );
      expect(wrapper.find('.col-6').contains(<p>Column Content</p>)).toBeTruthy();
    });

    test('Column renders with right alignment', () => {
      const wrapper = shallow(
        <Column right>
          <p>Column Content</p>
        </Column>,
      );

      expect(wrapper.find('.float-end').exists()).toBe(true);
    });
  });

  describe('Button Components', () => {
    const mockOnClick = jest.fn();

    test('Button.Success renders and handles click', () => {
      const wrapper = shallow(<Button.Success onClick={mockOnClick}>Success</Button.Success>);
      wrapper.find('button').simulate('click');
      expect(mockOnClick).toHaveBeenCalled();
      expect(wrapper.find('.btn-success').text()).toBe('Success');
    });

    test('Button.Danger renders and handles click', () => {
      const wrapper = shallow(<Button.Danger onClick={mockOnClick}>Danger</Button.Danger>);
      wrapper.find('button').simulate('click');
      expect(mockOnClick).toHaveBeenCalled();
      expect(wrapper.find('.btn-danger').text()).toBe('Danger');
    });

    test('Button.Light renders and handles click', () => {
      const wrapper = shallow(<Button.Light onClick={mockOnClick}>Light</Button.Light>);
      wrapper.find('button').simulate('click');
      expect(mockOnClick).toHaveBeenCalled();
      expect(wrapper.find('.btn-light').text()).toBe('Light');
    });
  });
});

describe('Row widget tests', () => {
  test('Draws correctly', () => {
    const wrapper = shallow(<Row>Row</Row>);

    expect(wrapper.matchesElement(<div className="row">Row</div>)).toEqual(true);
  });
});

describe('Column widget tests', () => {
  test('Draws correctly', () => {
    const wrapper = shallow(<Column>Content</Column>);

    expect(
      wrapper.matchesElement(
        <div className="col">
          <div className="float-start">Content</div>
        </div>,
      ),
    ).toEqual(true);
  });

  test('Draws correctly when property width is set', () => {
    const wrapper = shallow(<Column width={5}>content</Column>);

    expect(
      wrapper.matchesElement(
        <div className="col-5">
          <div className="float-start">content</div>
        </div>,
      ),
    ).toEqual(true);
  });

  test('Draws correctly when property right is set', () => {
    const wrapper = shallow(<Column right>content</Column>);

    expect(
      wrapper.matchesElement(
        <div className="col">
          <div className="float-end">content</div>
        </div>,
      ),
    ).toEqual(true);
  });
});

describe('Button.Success widget tests', () => {
  test('Draws correctly', () => {
    const wrapper = shallow(<Button.Success onClick={() => {}}>test</Button.Success>);

    expect(
      wrapper.matchesElement(
        <button type="button" className="btn btn-success">
          test
        </button>,
      ),
    ).toEqual(true);
  });

  test('Draws correctly with small button', () => {
    const wrapper = shallow(<Button.Success small onClick={() => {}}>test</Button.Success>);

    expect(
      wrapper.matchesElement(
        <button
          type="button"
          className="btn btn-success"
          style={{
            padding: '5px 5px',
            fontSize: '16px',
            lineHeight: '0.7',
          }}
        >
          test
        </button>,
      ),
    ).toEqual(true);
  });

  test('Button calls function on click-event', () => {
    let buttonClicked = false;
    const wrapper = shallow(
      <Button.Success onClick={() => (buttonClicked = true)}>test</Button.Success>,
    );

    wrapper.find('button').simulate('click');

    expect(buttonClicked).toEqual(true);
  });
});

describe('Button.Danger widget tests', () => {
  test('Draws correctly', () => {
    const wrapper = shallow(<Button.Danger onClick={() => {}}>test</Button.Danger>);

    expect(
      wrapper.matchesElement(
        <button type="button" className="btn btn-danger">
          test
        </button>,
      ),
    ).toEqual(true);
  });

  test('Draws correctly with small button', () => {
    const wrapper = shallow(<Button.Danger small onClick={() => {}}>test</Button.Danger>);

    expect(
      wrapper.matchesElement(
        <button
          type="button"
          className="btn btn-danger"
          style={{
            padding: '5px 5px',
            fontSize: '16px',
            lineHeight: '0.7',
          }}
        >
          test
        </button>,
      ),
    ).toEqual(true);
  });

  test('Button calls function on click-event', () => {
    let buttonClicked = false;
    const wrapper = shallow(
      <Button.Danger onClick={() => (buttonClicked = true)}>test</Button.Danger>,
    );

    wrapper.find('button').simulate('click');

    expect(buttonClicked).toEqual(true);
  });
});

describe('ButtonLight widget tests', () => {
  test('Draws correctly', () => {
    const wrapper = shallow(<Button.Light onClick={() => {}}>test</Button.Light>);

    expect(
      wrapper.matchesElement(
        <button type="button" className="btn btn-light">
          test
        </button>,
      ),
    ).toEqual(true);
  });

  test('Draws correctly with small button', () => {
    const wrapper = shallow(<Button.Light small onClick={() => {}}>test</Button.Light>);

    expect(
      wrapper.matchesElement(
        <button
          type="button"
          className="btn btn-light"
          style={{
            padding: '5px 5px',
            fontSize: '16px',
            lineHeight: '0.7',
          }}
        >
          test
        </button>,
      ),
    ).toEqual(true);
  });

  test('Button calls function on click-event', () => {
    let buttonClicked = false;
    const wrapper = shallow(
      <Button.Light onClick={() => (buttonClicked = true)}>test</Button.Light>,
    );

    wrapper.find('button').simulate('click');

    expect(buttonClicked).toEqual(true);
  });
});

describe('Alert tests', () => {
  test('No alerts initially', () => {
    const wrapper = shallow(<Alert />);

    expect(wrapper.matchesElement(<div></div>)).toEqual(true);
  });

  test('Show alert message', (done) => {
    const wrapper = shallow(<Alert />);

    Alert.danger('test');

    setTimeout(() => {
      expect(
        wrapper.matchesElement(
          <div>
            <div>
              test
              <button />
            </div>
          </div>,
        ),
      ).toEqual(true);

      done();
    });
  });

  test('Close alert message', (done) => {
    const wrapper = shallow(<Alert />);

    Alert.danger('test');

    setTimeout(() => {
      expect(
        wrapper.matchesElement(
          <div>
            <div>
              test
              <button />
            </div>
          </div>,
        ),
      ).toEqual(true);

      wrapper.find('button.btn-close').simulate('click');

      expect(wrapper.matchesElement(<div></div>)).toEqual(true);

      done();
    });
  });
});

describe('Details Component', () => {
  const mockProps = {
    title: 'Test Title',
    value: {
      id: 1,
      name: 'Test Name',
      content: '<p>Test Content</p>',
      birthDate: '1990-01-01',
      height: '180',
      country_name: 'Test Country',
      team_name: 'Test Team',
      pictureUrl: 'https://example.com/image.jpg',
      created_by: 'Test Creator',
      created_at: '2024-01-01',
      view_count: 123,
      revised_by: 'Test Reviser',
      revised_at: '2024-01-02',
      league_name: 'Test League',
      league_id: 2,
      team_id: 3,
      emblem_image_url: 'https://example.com/emblem.jpg',
      teams: [{ team_id: 1, team_name: 'Team A' }],
      players: [{ id: 1, name: 'Player A' }],
    },
    buttonText: 'Edit',
    onEdit: jest.fn(),
    onNavigate: jest.fn(),
    history: {
      length: 10,
      action: 'PUSH' as Action,
      location: { pathname: '', search: '', state: '', hash: '' },
      push: jest.fn(),
      replace: jest.fn(),
      go: jest.fn(),
      goBack: jest.fn(),
      goForward: jest.fn(),
      block: jest.fn(),
      createHref: jest.fn(),
      listen: jest.fn(),
    },
  };

  test('handles onNavigate for team and league links', () => {
    render(<Details {...mockProps} />);

    fireEvent.click(screen.getByText(mockProps.value.team_name));
    expect(mockProps.onNavigate).toHaveBeenCalledWith(`/team/${mockProps.value.team_id}`);

    fireEvent.click(screen.getByText(mockProps.value.league_name));
    expect(mockProps.onNavigate).toHaveBeenCalledWith(`/league/${mockProps.value.league_id}`);
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<Details {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
