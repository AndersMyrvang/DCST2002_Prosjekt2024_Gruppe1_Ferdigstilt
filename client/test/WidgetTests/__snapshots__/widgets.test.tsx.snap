// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="card card1"
  >
    <div
      class="card-body card-body1"
    >
      <h5
        class="card-title card-title1"
      >
        Test Title
      </h5>
      <div
        class="card-text card-text1"
      >
        <h2>
          Test Name
        </h2>
        <div
          class="row"
        >
          <div
            class="col-7"
          >
            <div
              class="float-start"
            >
              <div
                class="row"
              >
                <div
                  class="card card4 "
                >
                  <div
                    class="card-body card-body4 "
                  >
                    <h5
                      class="card-title card-title4"
                    />
                    <div
                      class="card-content card-content4"
                    >
                      <div>
                        <p>
                          Test Content
                        </p>
                      </div>
                      <div
                        class="row"
                      >
                        <div
                          class="col"
                        >
                          <div
                            class="float-start"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="col-3"
          >
            <div
              class="float-start"
            >
              <div
                class="row"
              >
                <div
                  class="card card2 "
                >
                  <div
                    class="card-body card-body2 "
                  >
                    <h5
                      class="card-title card-title2"
                    >
                      Test Name
                    </h5>
                    <div
                      class="card-content card-content2"
                    >
                      <div
                        class="col-11"
                      >
                        <div
                          class="float-start"
                        >
                          <div
                            class="card card4 "
                          >
                            <div
                              class="card-body card-body4 "
                            >
                              <h5
                                class="card-title card-title4"
                              />
                              <div
                                class="card-content card-content4"
                              >
                                <img
                                  alt="Image from https://example.com/emblem.jpg"
                                  src="https://example.com/emblem.jpg"
                                  width="360px"
                                />
                                <img
                                  alt="Image from https://example.com/image.jpg"
                                  src="https://example.com/image.jpg"
                                  width="360px"
                                />
                                <p>
                                  <b>
                                    Birth date:
                                  </b>
                                   1990-01-01
                                </p>
                                <p>
                                  <b>
                                    Height:
                                  </b>
                                   180 cm
                                </p>
                                <p>
                                  <b>
                                    Country:
                                  </b>
                                   Test Country
                                </p>
                                <p>
                                  <b>
                                    Team:
                                  </b>
                                   
                                  <a
                                    style="color: blue; text-decoration: underline; cursor: pointer;"
                                  >
                                    Test Team
                                  </a>
                                </p>
                                <p>
                                  <b>
                                    League:
                                  </b>
                                   
                                  <a
                                    style="color: blue; text-decoration: underline; cursor: pointer;"
                                  >
                                    Test League
                                  </a>
                                </p>
                                <p>
                                  <b>
                                    Teams:
                                  </b>
                                  <a
                                    style="color: blue; text-decoration: underline; cursor: pointer;"
                                  >
                                    <br />
                                    Team A
                                  </a>
                                </p>
                                <p>
                                  <b>
                                    Players:
                                  </b>
                                  <a
                                    style="color: blue; text-decoration: underline; cursor: pointer;"
                                  >
                                    <br />
                                    Player A
                                  </a>
                                </p>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="card card1"
        >
          <div
            class="card-body card-body1"
          >
            <h5
              class="card-title card-title1"
            >
              Page History
            </h5>
            <div
              class="card-text card-text1"
            >
              <div
                style="margin-left: 60px; display: flex; width: 1080px; justify-content: space-evenly; align-items: center;"
              >
                <div
                  style="display: flex; justify-content: space-evenly; width: 100%;"
                >
                  <p
                    style="flex: 1; text-align: left; margin-right: 20px;"
                  >
                    Laget av: Test Creator, 2024-01-01
                  </p>
                  <p
                    style="flex: 1; text-align: center; margin-right: 20px;"
                  >
                    Sist oppdatert av: Test Reviser, 2024-01-02
                  </p>
                  <p
                    style="flex: 1; text-align: right;"
                  >
                    View count: 123
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
